File: kotlin/src/main/kotlin/330.PatchingArray.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/patching-array/


class PatchingArray{
  fun minPatches(nums: IntArray, n: Int): Int {
    var ret = 0
    var cur = 0L
    var i = 0
    while(cur < n){
      if(i < nums.size && nums[i] <= cur+1)
        cur += nums[i++]
      else{
        ret ++
        cur += cur+1
      }
    }
    return ret
  }

  @Test
  fun test1(){
    assertEquals(1, minPatches(intArrayOf(1, 3), 6))
  }

  @Test
  fun test2(){
    assertEquals(2, minPatches(intArrayOf(1, 5, 10), 20))
  }

  @Test
  fun test3(){
    assertEquals(0, minPatches(intArrayOf(1, 2, 2), 5))
  }
}

File: python/330_PatchingArray.py
from unittest import TestCase
# https://leetcode.com/problems/patching-array/


class PatchingArray(TestCase):
    def minPatches(self, nums, n):
        """
        :type nums: List[int]
        :type n: int
        :rtype: int
        """
        x = i = ret = 0
        while x < n:
            if i < len(nums) and nums[i] <= x+1:
                x += nums[i]
                i += 1
            else:
                ret += 1
                x += x+1
        return ret

    def test1(self):
        self.assertEqual(1, self.minPatches([1, 3], 6))

    def test2(self):
        self.assertEqual(2, self.minPatches([1, 5, 10], 20))

    def test3(self):
        self.assertEqual(0, self.minPatches([1, 2, 2], 5))
